File: tyhp-lang/phpstorm/src/main/kotlin/com/tyhp/lang/binary/Platform.kt
package com.tyhp.lang.binary

import java.nio.file.Path

enum class OsId { OSX, LINUX, WIN }

enum class ArchId { X64, ARM64 }

enum class AssetVariant { SELF_CONTAINED, FRAMEWORK_DEPENDENT }

data class HostPlatform(
    val os: OsId,
    val arch: ArchId,
) {
    val osToken: String
        get() = when (os) {
            OsId.OSX -> "osx"
            OsId.LINUX -> "linux"
            OsId.WIN -> "win"
        }

    val archToken: String
        get() = when (arch) {
            ArchId.X64 -> "x64"
            ArchId.ARM64 -> "arm64"
        }
}

class UnsupportedPlatformError(
    val osName: String,
    val osArch: String,
) : RuntimeException(
    "Tyhp CLI has no GitHub Release asset for $osName/$osArch. " +
        "Supported assets: tyhp-osx-arm64, tyhp-osx-x64, tyhp-linux-x64, tyhp-linux-arm64, tyhp-win-x64.exe " +
        "(and matching -fxdependent variants).",
)

/**
 * GitHub Release asset names from `scripts/release.sh` EXPECTED_ASSETS.
 * Self-contained: `tyhp-{os}-{arch}` (Windows adds `.exe`).
 * Framework-dependent: same with `-fxdependent` before the optional `.exe`.
 */
fun releaseAssetName(platform: HostPlatform, variant: AssetVariant): String {
    val id = "${platform.osToken}-${platform.archToken}"
    val suffix = if (variant == AssetVariant.FRAMEWORK_DEPENDENT) "-fxdependent" else ""
    return if (platform.os == OsId.WIN) {
        "tyhp-$id$suffix.exe"
    } else {
        "tyhp-$id$suffix"
    }
}

fun installedBinaryFileName(platform: HostPlatform): String =
    if (platform.os == OsId.WIN) "tyhp.exe" else "tyhp"

fun pathProbeNames(platform: HostPlatform): List<String> =
    if (platform.os == OsId.WIN) listOf("tyhp.exe", "tyhp") else listOf("tyhp")

fun detectHostPlatform(
    osName: String = System.getProperty("os.name") ?: "",
    osArch: String = System.getProperty("os.arch") ?: "",
): HostPlatform {
    val osId = when {
        osName.contains("mac", ignoreCase = true) || osName.contains("darwin", ignoreCase = true) -> OsId.OSX
        osName.contains("linux", ignoreCase = true) -> OsId.LINUX
        osName.contains("win", ignoreCase = true) -> OsId.WIN
        else -> throw UnsupportedPlatformError(osName, osArch)
    }
    val arch = when (osArch.lowercase()) {
        "x64", "amd64", "x86_64" -> ArchId.X64
        "arm64", "aarch64" -> ArchId.ARM64
        else -> throw UnsupportedPlatformError(osName, osArch)
    }
    if (osId == OsId.WIN && arch == ArchId.ARM64) {
        throw UnsupportedPlatformError(osName, osArch)
    }
    return HostPlatform(osId, arch)
}

/** Matches `scripts/install.sh` (`$HOME/.local/bin`) and `scripts/install.ps1` (`%LOCALAPPDATA%\Programs\tyhp`). */
fun globalInstallDir(
    platform: HostPlatform,
    homedir: String,
    localAppData: String? = null,
): Path {
    if (platform.os == OsId.WIN) {
        val root = if (!localAppData.isNullOrBlank()) {
            localAppData
        } else {
            Path.of(homedir, "AppData", "Local").toString()
        }
        return Path.of(root, "Programs", "tyhp")
    }
    return Path.of(homedir, ".local", "bin")
}

fun globalInstallPath(
    platform: HostPlatform,
    homedir: String,
    localAppData: String? = null,
): Path = globalInstallDir(platform, homedir, localAppData).resolve(installedBinaryFileName(platform))

fun pluginInstallDir(pluginStorageFsPath: String): Path = Path.of(pluginStorageFsPath, "cli")

fun pluginInstallPath(pluginStorageFsPath: String, platform: HostPlatform): Path =
    pluginInstallDir(pluginStorageFsPath).resolve(installedBinaryFileName(platform))

/**
 * Whether a configured `tyhp.path` currently resolves to this plugin's managed
 * install location. `tyhp.binary.installMode` can drift from `tyhp.path` when a
 * user edits the path by hand (bypassing Install / Update CLI or a PATH re-probe);
 * auto-update must not treat a stale `installMode: "extension"` as license to
 * overwrite a path the user pointed elsewhere. Empty/unset paths are never managed.
 */
fun isManagedInstallPath(
    configuredPath: String,
    pluginStorageFsPath: String,
    platform: HostPlatform,
): Boolean {
    val trimmed = configuredPath.trim()
    if (trimmed.isEmpty()) {
        return false
    }
    val configured = Path.of(trimmed).toAbsolutePath().normalize()
    val managed = pluginInstallPath(pluginStorageFsPath, platform).toAbsolutePath().normalize()
    return configured == managed
}

/**
 * Plugin-only installs always use the self-contained asset so the IDE does not
 * depend on a machine-wide .NET 9 runtime. Global installs match `scripts/install.sh`:
 * framework-dependent when .NET 9 is present, otherwise self-contained.
 */
fun chooseAssetVariant(mode: String, hasDotNet9: Boolean): AssetVariant {
    if (mode == "extension") {
        return AssetVariant.SELF_CONTAINED
    }
    return if (hasDotNet9) AssetVariant.FRAMEWORK_DEPENDENT else AssetVariant.SELF_CONTAINED
}

File: tyhp-lang/vscode/src/binary/platform.ts
import * as os from "os";
import * as path from "path";

export type AssetVariant = "self-contained" | "framework-dependent";

export interface HostPlatform {
    os: "osx" | "linux" | "win";
    arch: "x64" | "arm64";
    nodePlatform: NodeJS.Platform;
}

export class UnsupportedPlatformError extends Error {
    constructor(readonly nodePlatform: string, readonly nodeArch: string) {
        super(
            `Tyhp CLI has no GitHub Release asset for ${nodePlatform}/${nodeArch}. ` +
                "Supported assets: tyhp-osx-arm64, tyhp-osx-x64, tyhp-linux-x64, tyhp-linux-arm64, tyhp-win-x64.exe " +
                "(and matching -fxdependent variants)."
        );
        this.name = "UnsupportedPlatformError";
    }
}

export function detectHostPlatform(
    nodePlatform: NodeJS.Platform = process.platform,
    nodeArch: string = process.arch
): HostPlatform {
    let osId: HostPlatform["os"];
    if (nodePlatform === "darwin") {
        osId = "osx";
    } else if (nodePlatform === "linux") {
        osId = "linux";
    } else if (nodePlatform === "win32") {
        osId = "win";
    } else {
        throw new UnsupportedPlatformError(nodePlatform, nodeArch);
    }

    let arch: HostPlatform["arch"];
    if (nodeArch === "x64" || nodeArch === "amd64") {
        arch = "x64";
    } else if (nodeArch === "arm64") {
        arch = "arm64";
    } else {
        throw new UnsupportedPlatformError(nodePlatform, nodeArch);
    }

    if (osId === "win" && arch === "arm64") {
        throw new UnsupportedPlatformError(nodePlatform, nodeArch);
    }

    return { os: osId, arch, nodePlatform };
}

/**
 * GitHub Release asset names from `scripts/release.sh` EXPECTED_ASSETS.
 * Self-contained: `tyhp-{os}-{arch}` (Windows adds `.exe`).
 * Framework-dependent: same with `-fxdependent` before the optional `.exe`.
 */
export function releaseAssetName(platform: HostPlatform, variant: AssetVariant): string {
    const id = `${platform.os}-${platform.arch}`;
    const suffix = variant === "framework-dependent" ? "-fxdependent" : "";
    if (platform.os === "win") {
        return `tyhp-${id}${suffix}.exe`;
    }
    return `tyhp-${id}${suffix}`;
}

export function installedBinaryFileName(platform: HostPlatform = detectHostPlatform()): string {
    return platform.os === "win" ? "tyhp.exe" : "tyhp";
}

export function pathProbeNames(platform: HostPlatform = detectHostPlatform()): string[] {
    if (platform.os === "win") {
        return ["tyhp.exe", "tyhp"];
    }
    return ["tyhp"];
}

/** Matches `scripts/install.sh` (`$HOME/.local/bin`) and `scripts/install.ps1` (`%LOCALAPPDATA%\Programs\tyhp`). */
export function globalInstallDir(
    platform: HostPlatform = detectHostPlatform(),
    homedir: string = os.homedir(),
    localAppData: string | undefined = process.env.LOCALAPPDATA
): string {
    if (platform.os === "win") {
        const root =
            localAppData && localAppData.trim() !== ""
                ? localAppData
                : path.join(homedir, "AppData", "Local");
        return path.join(root, "Programs", "tyhp");
    }
    return path.join(homedir, ".local", "bin");
}

export function globalInstallPath(
    platform: HostPlatform = detectHostPlatform(),
    homedir?: string,
    localAppData?: string
): string {
    return path.join(globalInstallDir(platform, homedir, localAppData), installedBinaryFileName(platform));
}

export function extensionInstallDir(globalStorageFsPath: string): string {
    return path.join(globalStorageFsPath, "cli");
}

export function extensionInstallPath(
    globalStorageFsPath: string,
    platform: HostPlatform = detectHostPlatform()
): string {
    return path.join(extensionInstallDir(globalStorageFsPath), installedBinaryFileName(platform));
}

/**
 * Whether a configured `tyhp.path` value currently resolves to this extension's own
 * managed install location. `tyhp.binary.installMode` can drift from `tyhp.path` when a
 * user hand-edits settings.json directly (bypassing "Tyhp: Install / Update CLI" or a PATH
 * re-probe, either of which resets `installMode`); auto-update must not treat a stale
 * `installMode: "extension"` as license to overwrite a `tyhp.path` the user pointed
 * elsewhere. Empty/unset paths are never considered managed.
 */
export function isManagedInstallPath(
    configuredPath: string,
    globalStorageFsPath: string,
    platform: HostPlatform = detectHostPlatform()
): boolean {
    const trimmed = configuredPath.trim();
    if (trimmed === "") {
        return false;
    }
    return path.resolve(trimmed) === path.resolve(extensionInstallPath(globalStorageFsPath, platform));
}

/**
 * Extension-only installs always use the self-contained asset so the IDE does not
 * depend on a machine-wide .NET 9 runtime. Global installs match `scripts/install.sh`:
 * framework-dependent when .NET 9 is present, otherwise self-contained.
 */
export function chooseAssetVariant(
    mode: "global" | "extension",
    hasDotNet9: boolean
): AssetVariant {
    if (mode === "extension") {
        return "self-contained";
    }
    return hasDotNet9 ? "framework-dependent" : "self-contained";
}
